runtime/memory/lazy.c

```diff
@@ -0,0 +1,143 @@
+#include "../runtime.h"
+
+#ifndef WASM_BLOCK_SIZE
+#define WASM_BLOCK_SIZE 4096
+#endif
+
+void **memory_blocks;
+u32 memory_size;
+
+void alloc_linear_memory() {
+    memory_blocks = malloc((starting_pages * WASM_PAGE_SIZE / WASM_BLOCK_SIZE)
+            * sizeof(void*));
+    memset(memory_blocks, 0, (starting_pages * WASM_PAGE_SIZE / WASM_BLOCK_SIZE)
+            * sizeof(void*));
+    memory_size = starting_pages * WASM_PAGE_SIZE;
+    printf("mem size: %d\n", memory_size);
+}
+
+void expand_memory() {
+    printf("mem size: %d -> %d\n", memory_size, memory_size + WASM_PAGE_SIZE);
+    // max_pages = 0 => no limit
+    assert(max_pages == 0 || (memory_size / WASM_PAGE_SIZE < max_pages));
+
+    memory_blocks = realloc(memory_blocks,
+            ((memory_size + WASM_PAGE_SIZE) / WASM_BLOCK_SIZE) * sizeof(void*));
+    assert(memory_blocks);
+
+    memory_blocks[memory_size / WASM_BLOCK_SIZE] = NULL;
+    memory_size += WASM_PAGE_SIZE;
+}
+
+// FIXME: This can't be provided by this memory backend
+//INLINE char* get_memory_ptr_for_runtime(u32 offset, u32 bounds_check) {
+//    assert(memory_size > bounds_check && offset <= memory_size - bounds_check);
+//
+//    char* mem_as_chars = (char *) memory;
+//    char* address = &mem_as_chars[offset];
+//
+//    return address;
+//}
```

Comment:
Is this function not used by the runtime? / How does the runtime compile without this function?